### DISCUSSION TITLE
Branch Protection on multiple branches

### DISCUSSION THREAD

**Discussion by clphillips** [community/community#22037] [New to GitHub] [answered]
<p>I’m wondering if there’s a way to apply branch protections in a regex kind of way. Here’s an example, I have multiple branches like <strong>release-1.0</strong> , <strong>release-1.1</strong>  and I’d like them to all have the same branch protections. More importantly, when I create a new <strong>release-2.0</strong> I’d like it to automatically be given branch protection with the same configuration as <strong>release-1.0</strong>  and <strong>release-1.1</strong>.</p>
<p>One way I’ve done this in other SCM management tools is by specifying a regular expression. Something to the effect of /^release-/</p>
<p>Is this possible?</p>

**Comment by ThirdEyeRose** [comment]
<p>Hi <a class="mention" href="/clphillips">@clphillips</a>,</p>
<p>Currently it is not possible to apply branch protections automatically in that way. However, we are looking to improve this feature over time, so I will pass on your feedback.</p>
<p>In the meantime, you can use the <a href="https://developer.github.com/v3/repos/branches/#update-branch-protection" rel="nofollow noopener">protected branches API endpoint</a> to script some automatic branch protection based on the branch name.</p>
<p>Hope this helps!</p>

**Comment by sfynx** [comment]
<p>It looks like it has been introduced this weekend. I can now create a protection with wildcards for something like release/* and it will automatically show all branches it affects, presumably also all future ones. Long overdue feature as other repo management systems had this for a while longer already, I’m glad it is here now.</p>

**Comment by OPeyrusse** [comment]
<p>Hello Sfynx,</p>
<p>seeing this new feature, I tried the same approach as you. However despite having many release branches, the branch restriction page is saying that “Applies to <strong>0</strong> branches”.</p>
<p>Must I do something special to make it work ? As suggested I set the pattern to “release/*”.</p>
<p>Many thanks</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/opeyrusse">@opeyrusse</a>,</p>
<p>Branch protection rule patterns are based on fnmatch syntax. You could use <code>releases/v?.?</code> to automatically protect branches like <code>releases/v1.0</code>, <code>releases/v2.0</code>, and <code>releases/v2.1</code>. And <code>[1-9]-[0-9]-stable</code> could automatically protect branches like <code>1-0-stable</code>, <code>2-0-stable</code>, and <code>2-1-stable</code>. To learn more about branch protection checkout <a href="https://help.github.com/articles/configuring-protected-branches/" rel="nofollow noopener">configuring protected branches.</a></p>
<p>I hope this helps!</p>
<p>Best,</p>
<p>Andrea</p>

**Comment by OPeyrusse** [comment]
<p>Thanks <a class="mention" href="/andreagriffiths11">@andreagriffiths11</a> for your answer.</p>
<p>It appears that <code>/</code> are treated specially. For me, the pattern <code>releases*</code> was not matching “releases/v1” but <code>releases/*</code> or any of your suggestions were working.</p>
<p>I also tested this on longer branch names, like “hotfix/client/project/1.7” and only the pattern <code>hotfix/*/*/*</code> was working. Removing intermediate “/” was breaking the pattern. It puzzled me as <a href="https://ruby-doc.org/core-2.5.1/File.html#method-c-fnmatch" rel="nofollow noopener">fnmatch documentation</a> and its examples seem to imply that a standard <code>hotfix/*</code> or at least <code>hotfix/**</code> should have worked.</p>
<p>Thanks a lot for your help.</p>

**Comment by ksraju-rms** [comment]
<p>I am not able to use github api to add branch protection for a branch that does not exists or a branch with wildcard. </p>
<p>Wanted to automate adding restrictions for all release branches (release/*) for all repos. Seems it does not work today and its a pain to add it manually as we have 100s of repos.</p>
<p>Ex: </p>
<p><a href="https://api.github.com/repos/:owner/:repo/branches/release/*/protection" rel="nofollow noopener">https://api.github.com/repos/:owner/:repo/branches/release/*/protection</a></p>
<p>Is there a way to do this automagically :slight_smile: ?? </p>

**Comment by nadiajoyce** [comment]
<p>That’s not possible at this time. However, I will share your feedback with the appropriate teams. I can see how this would be valuable! </p>

**Comment by trusktr** [comment]
<p>This would be great. Please!</p>
<p>I read in the another post that <code>{master,develop}</code> is not enabled in fnmatch. Can you provide another mechanism? Perhaps a UI with checkboxes as an alternative to pattern matching? That would be great!</p>

**Reply by dominictobias-bullish** [reply]
Everyone talking about pattern matching and this is all I want, you'd think they would give this most common example even in a hint on the UI :/

**Reply by patrick-knight** [reply]
[Repo rules](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/about-rulesets) has the native ability via the API and UI [to include multiple patterns per ruleset](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/creating-rulesets-for-a-repository#choosing-which-branches-or-tags-to-target)

**Comment by jeffaburt** [comment]
<p>+1 on adding support for this.</p>
<p>Looking more at <a href="https://ruby-doc.org/core-2.5.1/File.html#method-c-fnmatch" rel="nofollow noopener">the fnmatch documentation</a>, it looks like there’s a File::FNM_EXTGLOB option that can be passed to enable<span>{master,develop} </span>to work. This option is disabled by default. Enabling it would save so much time for our organization when new feature branches are created that multiple people will work off of.</p>

**Comment by Anahkiasen** [comment]
<p>Would love this as well</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/jeffaburt">@jeffaburt</a> and <a class="mention" href="/trusktr">@trusktr</a>,</p>
<p>Thanks for being here, we actually  have an internal issue discussing this possibility and I’ve added your feedback to it.  Thanks again for reaching out! I’ll come back and post any updates here.</p>

**Comment by tadms** [comment]
<p><a class="mention" href="/andreagriffiths11">@andreagriffiths11</a> </p>
<p>Another +1 for getting this enabled.  </p>
<p>I am preparing a migration of 100+ repos to GitHub Enterprise.  The inability to automate (via REST API) branch protection rules for wildcard patterns (such as release/* and feature/*) is going to be major painpoint for us.  </p>

**Comment by so-jelly** [comment]
<p><a class="mention" href="/tadms">@tadms</a>  I think the wildcard patterns support is already available and documented here: <a href="https://help.github.com/en/articles/configuring-protected-branches" rel="nofollow noopener">https://help.github.com/en/articles/configuring-protected-branches</a></p>
<p>The request still pending here is for multiple name matches such as a regex OR.   I supposed you could be looking for this?  {release/*,feature/*}   Would this also be supported?</p>

**Comment by tadms** [comment]
<p>Yes, this is possible through the UI.  But I’m not looking to manually do this 100+ times through the UI. </p>
<p>My request is to have this functionality with the REST API to be able to do this programmatically (originally requested by <a href="/t5/user/viewprofilepage/user-id/29066">ksraju-rms</a> in this thread).  </p>

**Comment by Zutatensuppe** [comment]
<p>Hello! <a class="mention" href="/andreagriffiths11">@andreagriffiths11</a> :)</p>
<p>What is the current status of that feature? Where can it be discussed? It would be awesome to be able to use a pattern like <code>{master,release/*,v[0-9]*}</code>.</p>

**Comment by AndreaGriffiths11** [comment]
<p>Thanks for the ping <a class="mention" href="/zutatensuppe">@zutatensuppe</a>, unfortunately I don’t have any updates for you “yet”</p>

**Comment by fishercraigj** [comment]
<p>I’d really like to be able to apply a single rule to the whole project as well no matter the branch name.</p>

**Comment by FelipeCortez** [comment]
<p><a class="mention" href="/fishercraigj">@fishercraigj</a>  Doesn’t <code>*</code> work?</p>

**Comment by fishercraigj** [comment]
<p>Not for me.  </p>

**Comment by joelpro2** [comment]
Reviving this thread
@AndreaGriffiths11  what's the status regarding to multiple root branches?
In my case, ```master``` and ```staging``` would have the exact same rules so i would expect something like ```{master,staging}```

**Reply by behitesh** [reply]
@AndreaGriffiths11  any luck with this... thank you.

**Reply by mtmelvin** [reply]
It kills me that nearly 5 years in, we still can't easily apply the same set of rules to multiple branches like this. Wildcards don't work in this case since the names are completely different.

**Reply by tylerlaprade** [reply]
I am interested in this as well.

**Reply by ysma500** [reply]
![image](https://user-images.githubusercontent.com/32311542/218553213-309222df-d38e-481e-8aed-4f773add747f.png)
Yeah, still bringing pain for a simple change: GitHub doesn't seem to pass the required File::FNM_EXTGLOB for this to actually work. Src. [Stack Overflow](https://stackoverflow.com/a/65438585)

**Reply by EstebanMongui** [reply]
Based on @ysma500 comment, here is the rule to match _develop_ and _main_ at the same rule 

`[md][ae][iv]*[np]`

**Note:** this rule will also apply for all branch's names which starts with 'm' or 'd' follows the sequence a-i, e-i, a-v or e-v and ends with 'n' or 'p'

**Comment by pohlt** [comment]
Fast-forward almost five(!) years and still no better solution than this `[ab][cd][ef]*` mess? Oh, please give us `EXTGLOB`. 🙏🏼

**Comment by Nauxscript** [comment]
there has been no progress in this issue for five years ... 🙃

**Comment by jonagh** [comment]
@AndreaGriffiths11 this discussion is set as answered, and I suppose it is as far as OP's question, however most of this discussion pertains to the fnmatch FNM_EXTGLOB feature being enabled so we can do something like "{branchone,branchtwo}". This seems like a very simple change and it has been years without action or indication. I'm concerned this issue is not being taken seriously because the discussion is marked as answered? or it's not being brought up with the relevant team? I don't know, but it's very bothersome to maintain a bunch of branch rules when one would do.

**Comment by patrick-knight** [comment, marked as answer]
Hey y'all,

Wanted to follow up on the discussion here, we recently introduced a beta of [repository rules](https://github.blog/changelog/2023-04-17-introducing-repository-rules-public-beta/). One feature is it does allow for defining multiple branch or tag patterns per ruleset. It's still not RegEx or `FNM_EXTGLOB` but via the UI and APIs you can define multiple patterns for the same set of rules. 

UI example:
<img width="785" alt="GitHubScreenShot-Arc-03-07-2023-001416@2x" src="https://github.com/community/community/assets/7575792/55ff97eb-1c80-4063-9716-6d66fcd5172b">


[API example:](https://docs.github.com/en/rest/repos/rules?apiVersion=2022-11-28#create-a-repository-ruleset)
```
  "conditions": {
    "ref_name": {
      "include": [
        "refs/heads/main",
        "refs/heads/master"
      ],
      "exclude": [
        "refs/heads/dev*"
      ]
    }
```

**Reply by patrick-knight** [reply]
Thanks to those who participated and shared solutions. As of the release of repository rules, this solution I posted above is the best way to manage this, as result we've unmarked the previous answer and updated the answer to reflect the best solution.

**Reply by gdostie** [reply]
Rulesets don't work for protecting [deployment branches](https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-deployments/managing-environments-for-deployment#deployment-branches-and-tags) thought.

**Reply by scarf005** [reply]
may I ask [why doesn't github branch rules support `FNM_EXTGLOB`](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-rulesets/creating-rulesets-for-a-repository#unsupported-fnmatch-syntax)? while `branch/{a,b}` situation could be workaround with adding multiple entries, it wouldn't be possible to support more complex cases such as `branch/{a,b}/name/{c,d}`.

![image](https://github.com/user-attachments/assets/9901c578-929f-49a6-902b-aa07d73957bd)
